rflx/statement.py

```diff
@@ -1,98 +1,151 @@
+    @abstractmethod
+    def check_type(
+        self, statement_type: rty.Type, typify_variable: Callable[[Expr], Expr]
+    ) -> RecordFluxError:
+        """Set the types of variables, and check the types of the statement and expressions."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def variables(self) -> Sequence[Variable]:
         raise NotImplementedError
 
 
 class Assignment(Statement):
-    def __init__(self, name: StrID, expression: Expr, location: Location = None) -> None:
-        super().__init__(name, location)
+    def __init__(self, identifier: StrID, expression: Expr, location: Location = None) -> None:
+        super().__init__(identifier, location)
         self.expression = expression
 
     def __str__(self) -> str:
-        return f"{self.name} := {self.expression}"
-
-    def validate(self, declarations: Mapping[ID, Declaration]) -> None:
-        if self.name not in declarations:
-            self.error.append(
-                f'assignment to undeclared variable "{self.name}"',
-                Subsystem.MODEL,
-                Severity.ERROR,
-                self.location,
-            )
-        else:
-            declarations[self.name].reference()
-        try:
-            self.expression.validate(declarations)
-        except RecordFluxError as e:
-            self.error.extend(e)
-        self.error.propagate()
+        return f"{self.identifier} := {self.expression}"
 
+    def check_type(
+        self, statement_type: rty.Type, typify_variable: Callable[[Expr], Expr]
+    ) -> RecordFluxError:
+        self.expression = self.expression.substituted(typify_variable)
+        return rty.check_type_instance(
+            statement_type, rty.Any, self.location, f'variable "{self.identifier}"'
+        ) + self.expression.check_type(statement_type)
 
-class Erase(Statement):
-    def __str__(self) -> str:
-        return f"{self.name} := null"
-
-    def validate(self, declarations: Mapping[ID, Declaration]) -> None:
-        if self.name not in declarations:
-            fail(
-                f'erasure of undeclared variable "{self.name}"',
-                Subsystem.MODEL,
-                Severity.ERROR,
-                self.location,
-            )
-        declarations[self.name].reference()
+    def variables(self) -> Sequence[Variable]:
+        return [Variable(self.identifier), *self.expression.variables()]
 
 
 class AttributeStatement(Statement):
     def __init__(
-        self, name: StrID, attribute: str, parameters: Sequence[Expr], location: Location = None
+        self,
+        identifier: StrID,
+        attribute: str,
+        parameters: Sequence[Expr],
+        location: Location = None,
     ) -> None:
-        super().__init__(name, location)
+        super().__init__(identifier, location)
         self.attribute = attribute
         self.parameters = parameters
 
     def __str__(self) -> str:
         parameters = ", ".join(str(p) for p in self.parameters)
```

Comment:
`join` is usually slightly faster if a list comprehension is used, see https://stackoverflow.com/a/37782238/4790678.